File: src/main/java/cz/vhromada/catalog/repository/GenreRepository.java
package cz.vhromada.catalog.repository;

import java.util.List;

import cz.vhromada.catalog.domain.Genre;
import cz.vhromada.common.repository.MovableRepository;

import org.springframework.data.jpa.repository.JpaRepository;
import org.springframework.util.CollectionUtils;

/**
 * An interface represents repository for genres.
 *
 * @author Vladimir Hromada
 */
public interface GenreRepository extends MovableRepository<Genre>, JpaRepository<Genre, Integer> {

    @Override
    default List<Genre> getAll() {
        return findAll();
    }

    @Override
    default Genre add(final Genre data) {
        assertGenre(data);

        return save(data);
    }

    @Override
    default Genre update(final Genre data) {
        assertGenre(data);

        return save(data);
    }

    @Override
    default List<Genre> updateAll(final List<Genre> data) {
        if (CollectionUtils.isEmpty(data)) {
            throw new IllegalArgumentException("Data mustn't be null or empty.");
        }

        return saveAll(data);
    }

    @Override
    default void remove(final Genre data) {
        assertGenre(data);

        delete(data);
    }

    @Override
    default void removeAll() {
        deleteAll();
    }

    /**
     * Checks genre
     *
     * @param genre genre
     * @throws IllegalArgumentException if genre is null
     */
    private static void assertGenre(final Genre genre) {
        if (genre == null) {
            throw new IllegalArgumentException("Genre mustn't be null.");
        }
    }

}

File: src/main/kotlin/cz/vhromada/catalog/repository/GenreRepository.kt
package cz.vhromada.catalog.repository

import cz.vhromada.catalog.domain.Genre
import org.springframework.data.jpa.repository.JpaRepository

/**
 * An interface represents repository for genres.
 *
 * @author Vladimir Hromada
 */
interface GenreRepository : JpaRepository<Genre, Int> {

    /**
     * Returns all genres created by user.
     *
     * @param user user's ID
     * @return all genres created by user
     */
    fun findByAuditCreatedUser(user: Int): List<Genre>

}
